//
// Copyright 2020 the authors listed in CONTRIBUTORS.md
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
syntax = "proto3";

package psi_proto;
option go_package = "github.com/openmined/psi/pb";

// Setup phase message for server.
message ServerSetup {
  message RawInfo {
    repeated bytes encrypted_elements = 1;
  }

  message GCSInfo {
    int32 div = 1;
    int64 hash_range = 2;
    bytes bits = 3;
  }

  message BloomFilterInfo {
    int32 num_hash_functions = 1;
    bytes bits = 2;
  }

  oneof data_structure {
    RawInfo raw = 1;
    GCSInfo gcs = 2;
    BloomFilterInfo bloom_filter = 3;
  }

}

// Client request with encoded elements sent to the server as an array of
// binary strings, together with a boolean reveal_intersection that
// indicates whether the client wants to learn the elements in
// the intersection or only its size.
message Request {
  bool reveal_intersection = 1;
  repeated bytes encrypted_elements = 2;
}

// Server response after encrypting client elements under the
// commutative encryption scheme, sent back to the client
// as an array of binary strings.
message Response {
  repeated bytes encrypted_elements = 1;
}
